src/pint/models/wavex.py

```diff
@@ -0,0 +1,387 @@
+                    units="s",
+                    value=wxcos,
+                    parameter_type="float",
+                    frozen=frozen,
+                )
+            )
+        self.setup()
+        self.validate()
+        return added_indices
+
+    def remove_wavex_component(self, index):
+        """
+        Remove all WaveX components associated with a given index or list of indices
+
+        Parameters
+        ----------
+        index : float, int, list, np.ndarray
+            Number or list/array of numbers corresponding to WaveX indices to be removed from model.
+        """
+
+        if isinstance(index, (int, float, np.int64)):
+            indices = [index]
+        elif isinstance(index, (list, set, np.ndarray)):
+            indices = index
+        else:
+            raise TypeError(
+                f"index most be a float, int, set, list, or array - not {type(index)}"
+            )
+        for index in indices:
+            index_rf = f"{int(index):04d}"
+            for prefix in ["WXFREQ_", "WXSIN_", "WXCOS_"]:
+                self.remove_param(prefix + index_rf)
+        self.validate()
+
+    def get_indices(self):
+        """
+        Returns an array of intergers corresponding to WaveX component parameters using WXFREQs
+
+        Returns
+        -------
+        inds : np.ndarray
+        Array of WaveX indices in model.
+        """
+        inds = [int(p.split("_")[-1]) for p in self.params if "WXFREQ_" in p]
+        return np.array(inds)
+
+    # Initialize setup
+    def setup(self):
+        super().setup()
+        # Get WaveX mapping and register WXSIN and WXCOS derivatives
+        for prefix_par in self.get_params_of_type("prefixParameter"):
+            if prefix_par.startswith("WXSIN_"):
+                self.register_deriv_funcs(self.d_wavex_delay_d_WXSIN, prefix_par)
+            if prefix_par.startswith("WXCOS_"):
+                self.register_deriv_funcs(self.d_wavex_delay_d_WXCOS, prefix_par)
+            self.wave_freqs = list(self.get_prefix_mapping_component("WXFREQ_").keys())
+            self.num_wave_freqs = len(self.wave_freqs)
+
+    def validate(self):
+        # Validate all the WaveX parameters
+        super().validate()
+        self.setup()
+        WXFREQ_mapping = self.get_prefix_mapping_component("WXFREQ_")
+        WXSIN_mapping = self.get_prefix_mapping_component("WXSIN_")
+        WXCOS_mapping = self.get_prefix_mapping_component("WXCOS_")
+        if WXFREQ_mapping.keys() != WXSIN_mapping.keys():
+            raise ValueError(
+                "WXFREQ_ parameters do not match WXSIN_ parameters."
+                "Please check your prefixed parameters"
+            )
+        if WXFREQ_mapping.keys() != WXCOS_mapping.keys():
+            raise ValueError(
+                "WXFREQ_ parameters do not match WXCOS_ parameters."
+                "Please check your prefixed parameters"
+            )
```

Comment:
Have you added tests to check that these exceptions are raised correctly?

Reply:
Added tests to check that exceptions are raised